### PR TITLE
complex shaper: avoid font warning for non-printable codepoints

### PR DESCRIPTION
**Pull request by avih**
This was performed in two steps (commits):
- Separate the non-printable simple-shaper detection into a stand-alone function.
- Reuse this function by the complex shaper.

### CODE REVIEW

**Comment by avih** [comment]
Repushed with trivial changes:
- Consistent termilonogy at the commit messages (non-printable/zero-width  -->  control-codepoint).
- Trivial change at the second commit: `if (is_control(info->symbol))` --> `if (is_control(symbol))`

**Comment by avih** [comment]
So there are two things which might need addressing:
- The second commit message states incorrectly that the simple shaper didn't show this warning - but it did, and the same commit addressed it unintentioally, as both shapers use `ass_shaper_find_runs` to resolve fonts. I'll revisit the commit message.
- The simple shaper now scans the same set of exclusion-codepoints twice: once when resolving the font, and once afterwards to mark it as skipped. Preferably we should only test each symbol once. I didn't try to evaluate yet where's the best place at the code for this would be. I think a good approach would be to scan and mark these symbols early, so that both shapers, and maybe also `ass_shaper_find_runs` would make use of the same scan result.

Or we can push it now, and unify the scans later - they do use the same code, but it does run twice now for the simple shaper. Let me know what you think.

**Comment by avih** [comment]
Re-pushed with these changes:
- First commit: only commit message and comments changes.
- Second commit: only commit message changes, and raised questions - hence marked as WIP.
- New third commit: avoid is_control redundancy for the simple shaper.

I'm copying the WIP questions asked at the commit message to allow easier discussion:
- What happens if harfbuzz does want a font for these codepoints?
- Should we use the same list of codepoint both to skip rendering at the simple shaper and to avoid font resolution? The codepoints are now documented, and it seems to me that the same list should be OK.
- What if none of the symbols resolves a font? (not a new issue).

Once we answer those, I'll remove the WIP text from the second commit.

And, the last commit message suggests that it might be possible to reuse the existing `.skip` member instead of adding a new `.is_control` member. It's probably fine as-is currently, and here are some considerations:
- They do seem at least somewhat overlapping in function.
- Reusing will definitely NOT add any regression for the complex shaper, because right after `ass_shaper_find_runs` (which tests `is_control`) - `ass_shaper_shape` is called, and at the complex shaper code path it resets and re-calculates the `.skip` member anyway (at `shape_harfbuzz`).
- For the simple shaper code path, it can cause regressions only if the `.skip` member is set selectively before calling `ass_shaper_find_runs`. If it's not, then immediately afterwards `shape_fribidi` is called (which doesn't seem to use `.skip`), and then `ass_shaper_skip_characters` (which sets it to the same values calculated earlier by `is_control`).

However, even if `.skip` can be reused 100% safely, there's still a question of whether we want to. I don't think I can answer this on my own, so for now I think it's OK to use the new member `.is_control`.

**Comment by avih** [comment]
> Yes, that sounds like a regression, unfortunately. Good catch!

Right. Good catch, bad patch :)

> _must_ make it to HarfBuzz as part of their surrounding shaping runs

Indeed. I also presumed this could be the case, though as you mentioned, it can happen anyway before this PR if the glyph is missing - which as we know can definitely happen for control codepoints.

> I think that’s what you do in this PR, so is the question rather “is there anything wrong with this approach”?

Correct. I thought this was obvious in this context, but I could have phrased it better. Thanks.

> ideally I’d like to somehow make HarfBuzz pick these characters out for us

Completely agreed. That was one of the things I thought might be happening here https://github.com/libass/libass/issues/507#issuecomment-834321785

It's definitely better to not guesswork/duplicate it, unless it's specified explicitly how harfbuzz does this.

> > What if none of the symbols resolves a font? (not a new issue).
> 
> Uh, well, is it really any different from when _some_ of the symbols don’t resolve any font?

I _think_ it is different. Because if no font was resolved, and later you're required to provide a font, then it might end up accessing index 0 of the array, and maybe the array size is 0 to begin with because no font was ever found.

I didn't try to follow the code enough to understand whether it's a real issue or not, but regardless, it's not a new issue because theoretically (but I don't know if practically) it can already happen before this PR if all the symbols are smaller than x20.

> So you can just move the `ass_shaper_skip_characters` call from the simple shaper path in `ass_shaper_shape` to the common path in `ass_shaper_find_runs` or `ass_render_event`.

Can I? I wouldn't have thought so myself, because then `shape_fribidi` - which runs after `ass_shaper_find_runs` would not see the direction codepoints (because by then the symbol was modified to 0).

So, considering the potential (additional) regression with even more cases of too-short runs than before, here's what I suggest we do:
1. Do this now https://github.com/libass/libass/issues/507#issuecomment-835368381 (just suppress warnings for control codepoints). PR-wise, this means keeping only the first commit, and adding a new commit which reuses this function to test whether it should warn or just do info.
2. At a later point someone might want to have a go at the more correct https://github.com/libass/libass/issues/507#issuecomment-835375915 (refactor the dataflow - which I accidentally thought was posted by you, and only now noticed it's by a different person).

**Comment by astiob** [comment]
> because by then the symbol was modified to 0

Er, right. Just drop the part that changes symbol to 0. (It’s important for the complex shaper, too.) I don’t think it adds any meaning for later processing stages anyway (when `skip` is set)!

**Comment by astiob** [comment]
On the other hand—not directly relevant to this PR, but—if the font does contain a glyph for a control character and we’re using the “simple shaper”, I _think_ we end up using the glyph’s metrics for layout despite `ass_shaper_skip_characters`. I’m not too confident though, so this needs to be tested.

**Comment by avih** [comment]
> Er, right. Just drop the part that changes symbol to 0. (It’s important for the complex shaper, too.) I don’t think it adds any meaning for later processing stages anyway (when `skip` is set)!

Right. That was indeed one thing I also considered, but didn't know if both `.skip` and `.symbol=0` are required, or whether `.skip` is enough. And that's also the reason I thought that `.skip` might be set selectively before `ass_shaper_find_runs` in a way which is not completely overlapping `.symbol==0` - and ultimately the reason I didn't reuse `.skip`.

> It’s probably safe to stick to copying the run ID of the preceding character and use 0 or `skip` or whatever if the control character itself leads the string.

Yes, definitely better to not change the run ID, but it can become tricky to a degree I felt uncomfortable to make this change without deeper knowledge of the codebase.

So, I still stand by this:

> So, considering the potential (additional) regression with even more cases of too-short runs than before, here's what I suggest we do:
> 
>  Do this now ... (just suppress warnings for control codepoints). PR-wise, this means keeping only the first commit, and adding a new commit which reuses this function to test whether it should warn or just do info.

Now also because different approaches which work better than this PR do also require deeper knowledge of the codebase, and I'm simply not comfortable taking them without knowing for a fact that I don't introduce regressions.

If you think that we can do better right now than simply addressing the warning on its own (i.e. my suggestion), then I'd prefer that you take over and handle it however you think is best.

What say you?

**Comment by avih** [comment]
> That means the control character must inherit `shape_run_id` of the preceding (or succeeding) character, regardless of whether and which fonts contain or lack a glyph for it and regardless of whether the preceding (or succeeding) character itself comes from the main font or a fallback font.

I agree that's the best thing to do, and I decided to have a go at this too. I think on top of the existing 3 commits would be OK.

> It’s probably safe to stick to copying the run ID of the preceding character and use 0 or `skip` or whatever if the control character itself leads the string.

That's implementation details, which I don't immediately know how to translate to code. Also, not sure about `.skip`. I recommend that for now, unless you think this would be incorrect, keep the existing usage of `skip` independent of the new code. This can be optimized later, instead of adding noise to the current discussion.

I've had a different(?) idea at the implementation:
- The base is the current code - with the 3 commits which are already at this PR.
- Keep the current function of `ass_shaper_find_runs` as is, but with an additional single bit which remembers whether at least one control codepoint was encountered throughout the entire length (spanning runs).
- After the loop is complete - if no control codepoints were encountered or if there was only one run (`shape_run == 0`), then return normally. I.e. so far this is identical as before - but only for cases where control codepoints didn't exist or didn't break the run.
- Otherwise - i.e. there were control codepoint[s] and more than one run: make a second pass of the whole length and re-calculate the runs. Depending on answers to the questions below - this might be integrated with the first pass instead of as an additional pass.

Few things I'd need help with:
- Does anyone care later on if a run has more than one font? Specifically - can we leave the font as 0 for control codepoints, while surrounding symbols of the same run have a different font? If no one cares then it will make the logic quite simpler, because we wouldn't need to find an actual font from another symbol at the run to copy to the control codepoint[s] of the run.
- Why is `ass_font_get_index` (which resolves a font) called only if `(!info->drawing_text.str)` ? I.e. what's the meaning of having or not having `info->drawing_text.str`? My guess would be that the code considers `info->drawing_text.str != NULL` if and only if font info was already resolved for this symbol. Is this correct?
- At the existing code, to test whether a new run starts, the font info is compared with the previous glyph at the array - NOT with the previous glyph for which `ass_font_get_index` was executed (e.g. if `info->drawing_text.str` is not NULL). To me it looks that the code assume (see previous question) that the glyph already has front info. Is this correct? who sets the font prior to running `ass_shaper_find_runs`? And why only to some of the symbols at the array?

**Comment by avih** [comment]
> Does anyone care later on if a run has more than one font?
> Assuming `info->drawing_text.str != NULL` indeed means "already has font info"

Assuming no one cares if control codepoints inside a run have a different font than surrounding symbols at the run, and assuming `info->drawing_text.str != NULL` indeed means "already has font info", then here's my suggested patch on top of the existing 3 commits. I.e. This integrates the new logic into the existing pass without adding a new pass:
```diff
diff --git a/libass/ass_shaper.c b/libass/ass_shaper.c
index 396185f..50390c6 100644
--- a/libass/ass_shaper.c
+++ b/libass/ass_shaper.c
@@ -835,8 +835,17 @@ void ass_shaper_find_runs(ASS_Shaper *shaper, ASS_Renderer *render_priv,
             ass_font_get_index(render_priv->fontselect, info->font,
                     symbol, &info->face_index, &info->glyph_index);
         }
-        if (i > 0) {
-            GlyphInfo *last = glyphs + i - 1;
+
+        // test if a new run starts, completely ignoring control codepoints
+        int last_font = info->is_control ? -1 : i - 1;
+        while (last_font >= 0 && glyphs[last_font].is_control)
+            last_font--;
+        // last_font is -1 if NOT considering a new run (first symbol, or
+        // control codepoint, or all prior symbols are control), else nearest
+        // earlier index to compare font (guaranteed not a control codepoint).
+
+        if (last_font >= 0) {
+            GlyphInfo *last = glyphs + last_font;
             if ((last->font != info->font ||
                     last->face_index != info->face_index ||
                     last->script != info->script ||
```

This fixes the new cases of too-short runs introduced by the second commit - control codepoint breaks the run even if a glyph can be found.

And it also fixes earlier cases of too-short runs - control codepoint breaks the run because no glyph was found.

**Comment by avih** [comment]
> I don’t quite understand what you mean

OK, I based my question on this (https://github.com/libass/libass/issues/507#issuecomment-834345013):
> ass_shaper_shape runs HarfBuzz, which requests font glyphs via callbacks and receives our previously found glyphs. I assume it doesn’t request a glyph for U+200B or U+2060. ...

So I'm trying to understand for a fact what happens without assuming: whether or not harfbuzz does try to "request the font glyph via callbacks" for codepoints at some default_ignorable list.

And the reason I'm asking is to understand whether or not libass should try to resolve a face which has a glyph for it, or whether it can simply say "it's ignorable, I'm not going to bother finding a (fallback) face which has this glyph".

The second question is basically: does harfbuzz care at all what the face is for codepoints in a default_ignorable list - under normal circumstances (without some "show hidden" mode etc).

I.e. I _think_ it's not unimaginable that harfbuzz knows that the glyph itself is not relevant under normal circumstances, but that the face might still be relevant to extract some metrics in order to apply the codepoint correctly during layout. This is a guess - but I hoped for more knowledgable info by you.

The reason I'm asking is that due to implementation details, if libass doesn't try to resolve a face for the glyph then the face ends up arbitrary. So this patchset can either leave it with some arbitraty value (if harfbuzz doesn't care what the face is) or it needs to do some work in order to adjust the face to match the resolved face of surrounding glyphs.

So this is how I interpret your reply (please confirm):
> But HarfBuzz will behave shaping wise whether the font have glyphs for these code points or not. Under normal circumstances it will replace them with space glyph and zero its widths (the replacement glyph can be controlled with the API)

So assuming a replacement glyph is not provided via the API, and under "normal" circumstances, do I understand correctly that harfbuzz does not need to access neither the glyph nor the font which is associated with this codepoint?

**Comment by astiob** [comment]
> the check is tailored for font shaping in HarfBuzz, it doesn't match the Unicode property 100%

Hmm, so basically, there’s no way to know which code points require font data and which don’t other than going the Chrome way (with extra complications due to synthesized glyphs) or duplicating internal HarfBuzz code?

Speaking of the Chrome way, is it not possible that a font intentionally uses glyph #0 for a supported character (or substitution) such that it should _not_ invoke font fallback?

And the proper handling of default ignorable code points is… feed them to HarfBuzz first; if it returns .notdef, perform font fallback as usual; if out of fonts and they’re still .notdef, turn them invisible? It would be great if hb-unicode exposed the raw default-ignorable property so we didn’t have to maintain our own list likely to fall out of sync over time.

(By the way, I don’t see Chrome do this—turn ignorables invisible if no font has them and HarfBuzz has let them through—although I’m merely foolishly looking at unfamiliar source code. It looks like Firefox does this. But then Firefox had [this issue](https://bugzilla.mozilla.org/show_bug.cgi?id=1238243) but I don’t understand how it could be an issue to begin with and hence whether I need to worry about it in libass.)

**Comment by astiob** [comment]
@avih Sorry for my silence in the last few days. I very much appreciate your enthusiasm and effort, but I think you’re overcomplicating this :-)

Just copy the last glyph’s font face in `ass_shaper_find_runs` instead of invoking `ass_font_get_index`.

By the way, currently, libass will try to fetch the glyph later anyway, regardless of HarfBuzz.

File: libass/ass_shaper.c [hunk 1]
```diff
@@ -800,26 +819,75 @@ void ass_shaper_find_runs(ASS_Shaper *shaper, ASS_Renderer *render_priv,
...
+            //       resolved by ass_font_get_index for other reasons
+            //       (glyph_index != 0 if and only if a face was resolved).
+            if (info->is_control)
+                face_comparable = false;
+
             // set size and get glyph index
+            // NOTE: if face_index changes then it can split the run further,
+            //       but cannot accidentally combine adjacent runs (e.g. if two
+            //       symbols had different face_index but they have the same
+            //       face_index after ass_font_get_index). that's because split
```
[the remark this answers is not in the document]
Comment (avih) [reply libass/ass_shaper.c:854]: > FYI: This is the very place that sets `face_index`. It is zero until here.

Right, that's the only thing I wasn't 100% sure of, so I added this comment to make it obvious if I'm assuming wrong. Thanks.

The reason I assumed that is that `split_style_runs` splits on bold and italic, and assumed `face_index` was derived from the style somewhere before `ass_shaper_find_runs`, and that then `ass_font_get_index` starts with the desired face/style, and then falls back to others. That's the very first thing which `ass_font_get_index` does - tries the glyph at the input face_index - and if it's found, then nothing happens further later:
```c
    // try with the requested face
    if (*face_index < font->n_faces) {
        face = font->faces[*face_index];
        index = FT_Get_Char_Index(face, ass_font_index_magic(face, symbol));
    }
```

**Comment by avih** [comment]
> > maybe the array size is 0 to begin with because no font was ever found.
> 
> This is impossible, so you need not worry ..

Good. Thanks :)

> I was about to say:
> But for almost all default-ignorable code points ...

Right. But I didn't make any assumptions about prior states, except at my comment which you say is wrong :) (about face_index always being 0 before calling `ass_font_get_index`). Luckily the comment is wrong only in its assumption that face_index could be different on entry, but not in its conclusion - that if `ass_font_get_index` changes face_index then it can only split the run further but not accidentally combine adjacent runs - which was my main concern.

TBH, by looking at `ass_font_get_index` it still definitely does _not_ look to me like a function which expects face_index to always be 0 on entry, and that n_faces is always at least 1, but I'll take your word that this is the case.

> HarfBuzz can keep one of these code points visible ...

I don't think I really understand the statement or its implications. See below.

> I think this alone actually answers all your questions about needing the right font.
> Oh welp, so we do want the right face_index for run-leading code points.

Not 100% sure I follow. The right font as the surrounding glyphs (i.e. artificial adjustment)? or the right font as one which has a glyph for this codepoint (i.e. we actually need `ass_font_get_index` to try and resolve a face rather than calling it with symbol==0)?

If it's the latter and it does have a resolved face which (can) differs from its surroundings, do we want to break the run due to the changed face (assuming nothing else except the face changed)? or do we want to keep it in one run which will now have more than one face?

And also if it's the latter, what happens if we can't find a face which has this glyph? Do we still do the warning?

> ASS can contain vector drawings instead of (in addition to) text. Drawings need no text shaping and delimit shaper runs.

Right. I didn't assume anything about it, and the code doesn't need modification following this new (to me) information, but it does answer one of the TODO items:
```
// TODO: else face_comparable remains true. is this correct?
```
So the answer is that it's correct because it doesn't matter - it's force-delimited by `starts_new_run` anyway. Good.

**Comment by khaledhosny** [comment]
> So I'm trying to understand for a fact what happens without assuming: whether or not harfbuzz does try to "request the font glyph via callbacks" for codepoints at some default_ignorable list.

It depends on whether HarfBuzz wants to show this glyph or not, and like I said earlier there are few situations where it would.

> And the reason I'm asking is to understand whether or not libass should try to resolve a face which has a glyph for it, or whether it can simply say "it's ignorable, I'm not going to bother finding a (fallback) face which has this glyph".

This should work, as long as it is kept in the same run (runs, by definition, must use the same font) with its surroundings.

> Speaking of the Chrome way, is it not possible that a font intentionally uses glyph #0 for a supported character (or substitution) such that it should _not_ invoke font fallback?

HarfBuzz does not have any special meaning of 0 glyph index, the callback function returns false when it does not find a glyph for the specified code point and that what tells HarfBuzz the font does not support it (Pango, for example, returns a very high glyph index that also encodes the code point to use it later for drawing missing glyph hex box).

> feed them to HarfBuzz first; if it returns .notdef, perform font fallback as usual; if out of fonts and they’re still .notdef, turn them invisible

You are unlikely to ever end up with `.notdef` for Default_Ignorables unless no font in the fallback stack has the replacement character (space by default).

**Comment by astiob** [comment]
> You are unlikely to ever end up with `.notdef` for Default_Ignorables unless no font in the fallback stack has the replacement character (space by default).

This is true for most of them, but what about the Default_Ignorables that are excluded from HarfBuzz’s internal list, such as hangul fillers and shorthand format controls? My understanding is that it does not treat them as ignorables in any manner at all. If the font supports them, the font itself will ligate/substitute them into blank glyphs. But if not, we should not show fallback .notdef, because that’s the definition of Default_Ignorable.

**Comment by astiob** [comment]
> HarfBuzz does not have any special meaning of 0 glyph index, the callback function returns false when it does not find a glyph for the specified code point and that what tells HarfBuzz the font does not support it (Pango, for example, returns a very high glyph index that also encodes the code point to use it later for drawing missing glyph hex box).

Aha, I see it now. Nice, thanks!

If the `hb_font_get_nominal/variation_glyph` returns false, is it safe to assume HarfBuzz will not call any of the other callbacks for that glyph?

**Comment by khaledhosny** [comment]
> This is true for most of them, but what about the Default_Ignorables that are excluded from HarfBuzz’s internal list

They would be returned as `.notdef` indeed, though I'm not familiar with these code points to know if it is an issue in practice.

> If the `hb_font_get_nominal/variation_glyph` returns false, is it safe to assume HarfBuzz will not call any of the other callbacks for that glyph?

I think so.

**Comment by avih** [comment]
Force-pushed two commits which replace the previous ones, as requested by @astiob on IRC.

**Comment by avih** [comment]
For what it's worth, I've been using the last patch-set (two-commits) in my own build of libass with mpv for the past 2 weeks. It seems to solve the original issue (no more warnings for missing glyphs of default-ignorable), and I didn't notice a negative impact, though I didn't try to measure anything.

**Comment by astiob** [comment]
I’ll get back to this now.

Preliminary tests suggested that Uniscribe in VSFilter _doesn’t_ attach leading ZWJ to a subsequent fallback shape run, which should let us simplify our code further. But first, I want to do another test with something like U+1D179 MUSICAL SYMBOL BEGIN PHRASE that is _logically_ a leading character, to make sure it doesn’t depend on the exact character.

**Comment by avih** [comment]
> Preliminary tests suggested ...

I don't think I understand how is this related to the issue which this PR tries to solve - avoid warning for missing glyphs which are not rendered anyway.

**Pull request closed without merging**